File: Examples/Ex5/Ex5Class.cpp
#include <TMB.hpp>

template <class Type> Type square(Type x){return x*x;}

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_INTEGER(m); //number of stocks
  DATA_IVECTOR(TT); //number of years
  DATA_INTEGER(Tmax); //max number of years
  DATA_MATRIX(B); //biomass matrix
  DATA_MATRIX(R); //recruitment matrix
  DATA_VECTOR(Phi0); //spawner biomass per recruit

  // End of data section

  PARAMETER(dummy); 
  PARAMETER(mu); //mean of random effect
  PARAMETER(log_tau); //log of sd of random effect
  PARAMETER_VECTOR(B0); //initial steepness; bounded between 0.1 and 1000
  PARAMETER_VECTOR(log_sigR); //i think this is log of sd of F
  PARAMETER_VECTOR(eta); //no bounds, part of steepness equation 
  // End of the estimated parameters

  vector<Type> R0(m); //stock-specific unfished recruitment; is B0/Phi0
  vector<Type> SigR(m); //this is sd on recruitment
  vector<Type> h(m); //steepness
  vector<Type> beta; //this is random effect of steepness
  matrix<Type> BH(m,Tmax); //this represents the first part of R equation; aka F*Bit 
  matrix<Type> eps(m,Tmax); //i created this to represent the epsilon in the main Rit equation
  
  Type tau; //sd of random effect
  Type obj_fun;
  Type NegLogLike = 0;
  
  // End of the temporary variables

  // Transform the parameters
  R0 = B0/Phi0;
  tau = exp(log_tau);
  SigR = exp(log_sigR);
  obj_fun = 0;
  
  // Extract beta and define h
  for (int i=0;i<m;i++) {
    beta[i] = log((h[i]-0.2)/(1-h[i]));
   for (int j=0;j<TT;j++) {
    BH[i,j] = (4 * R0[i] * h[i] * B[i,j]) / (Phi0[i] * R0[i] * (1-h[i]) + (((5 * h[i]) - 1) * B[i,j]));
    R[i,j] = BH[i,j] * exp(eps[i,j] - (SigR[i]*SigR[i]/2));
    eps[i,j] ~ dnorm(0,SigR);
    B[i,j] ~ dnorm(mu,tau);
   }}

  NegLogLike += (log(R) - log(BH) + ((SigR * SigR)/2)) * (log(R) - log(BH) + ((SigR * SigR)/2)) / ((2 * SigR * SigR) + log_sigR)
  
  obj_fun += dummy*dummy;
  ADREPORT(h);
  ADREPORT(R0);
  ADREPORT(tau);

  return(obj_fun);
}

File: Examples/Ex5/ex5Class.R

################################################################################

setwd("G:/My Drive/GitHub/TMB_Workshop_Sep2022/Examples/Ex5")
require(TMB)

################################################################################

m <- scan("ex5.dat",skip=1,n=1,quiet=T) #50; number of stocks/species
TT <- scan("ex5.dat",skip=3,n=m,quiet=T) #vector of 36 50x; number of years by stock
Tmax <- scan("ex5.dat",skip=5,n=1,quiet=T) #36; max # years
B <- matrix(scan("ex5.dat",skip=7,n=m*Tmax,quiet=T),ncol=Tmax,byrow=T) #matrix 50x36; spawning biomass
R <- matrix(scan("ex5.dat",skip=58,n=m*Tmax,quiet=T),ncol=Tmax,byrow=T) #matrix 50x36; recruitment at age 0
Phi0 <- scan("ex5.dat",skip=109,n=m,quiet=T); #vector of 50; spawner biomass-per-recruit wo exploitation for each stock

################################################################################
# Hint Nproj = 0 for basic estimation
# Basic estimation
################################################################################

# Set the parameters to initial values
data <- list(m=m,TT=TT,Tmax=Tmax,B=B,R=R,Phi0=Phi0)
parameters <- list(dummy=0,mu=mu,log_tau=log_tau,B0=B0,log_sigR=log_sigR,eta=eta)

#print(data)
#print(parameters)

compile("Ex5Class.cpp", flags="-Wno-ignored-attributes")
dyn.load(dynlib("Ex5Class"))

# test code - for checking for minimization
map<-list(mu=factor(NA),log_tau=factor(NA),B0=rep(factor(NA),m),log_sigR=rep(factor(NA),m),eta=rep(factor(NA),m))
model <- MakeADFun(data, parameters, DLL="Ex5",silent=T,map=map)
xx <- model$fn(model$env$last.par)
cat(xx,model$report()$obj_fun,"\n")
#AAA

# Now estimate everything
map<-list(dummy=factor(NA))
model <- MakeADFun(data, parameters, random="??", DLL="Ex5",silent=T,map=map)

# Bounds on the parameters
lowbnd=c(NA,NA,-7,0.1,-7,NA)
uppbnd=c(NA,NA,4,1000,4,NA)
fit <- nlminb(model$par, model$fn, model$gr, control=list(rel.tol=1e-12,eval.max=100000,iter.max=1000),
              lower=lowbnd,upper=uppbnd)
best <- model$env$last.par.best
print(best)
rep <- sdreport(model)
print(summary(rep))


